x/autopilot/types/parser.go

```diff
@@ -1,51 +1,132 @@
-	// TODO: Use JSON parsing instead of string parsing
-	parts := strings.Split(receiverData, "|")
-
-	switch len(parts) {
-	case 1:
-		return &ParsedReceiver{
-			ShouldLiquidStake: false,
-		}, nil
-	case 2:
-		addressPart := parts[0]
-		functionPart := parts[1]
-
-		// verify the rightmost field is stakeibc/LiquidStake
-		if functionPart != "stakeibc/LiquidStake" {
-			return &ParsedReceiver{
-				ShouldLiquidStake: false,
-			}, nil
-		}
-
-		strideAccAddress, err := sdk.AccAddressFromBech32(addressPart)
-		if err != nil {
-			return nil, err
-		}
-
-		return &ParsedReceiver{
-			ShouldLiquidStake: true,
-			StrideAccAddress:  strideAccAddress,
-		}, nil
-	default:
-		return nil, ErrInvalidReceiverData
+// QUESTION: Should we just leave the address as a string in the parsed type
+// and then combine these raw and parsed types?
+type rawPacketMetadata struct {
+	Autopilot *struct {
+		Stakeibc *RawStakeibcPacketMetadata `json:"stakeibc"`
+		Claim    *RawClaimPacketMetadata    `json:"claim"`
+	} `json:"autopilot"`
+}
+
+type RawStakeibcPacketMetadata struct {
+	Action        string `json:"action"`
+	StrideAddress string `json:"stride_address"`
+}
+
+type RawClaimPacketMetadata struct {
+	AirdropId     string `json:"airdrop_id"`
+	StrideAddress string `json:"stride_address"`
+}
+
+type PacketMetadata struct {
+	Reciever string
+	Stakeibc StakeibcPacketMetadata
+	Claim    ClaimPacketMetadata
+}
+
+type StakeibcPacketMetadata struct {
+	Enabled       bool
+	Action        string
+	StrideAddress sdk.AccAddress
+}
+
+type ClaimPacketMetadata struct {
+	Enabled       bool
+	AirdropId     string
+	StrideAddress sdk.AccAddress
+}
+
+// Validate stakeibc packet metadata fields
+func (r *RawStakeibcPacketMetadata) ParseAndValidate() (*StakeibcPacketMetadata, error) {
+	// If the stakeibc section of the memo field was empty, mark stakeibc as disabled
+	if r == nil {
+		return &StakeibcPacketMetadata{Enabled: false}, nil
+	}
+
+	// Validate the stride address and action
+	address, err := sdk.AccAddressFromBech32(r.StrideAddress)
+	if err != nil {
+		return nil, err
+	}
+	if r.Action != "LiquidStake" {
+		return nil, errorsmod.Wrapf(ErrUnsupportedStakeibcAction, "action %s is not supported", r.Action)
+	}
+
+	return &StakeibcPacketMetadata{
+		Action:        r.Action,
+		StrideAddress: address,
+		Enabled:       true,
+	}, nil
+}
+
+// Validate claim packet metadata fields
+func (r *RawClaimPacketMetadata) ParseAndValidate() (*ClaimPacketMetadata, error) {
+	// If the claim section of the memo field was empty, mark claim as disabled
+	if r == nil {
+		return &ClaimPacketMetadata{Enabled: false}, nil
+	}
+
+	// Validate the stride address and airdrop ID
+	address, err := sdk.AccAddressFromBech32(r.StrideAddress)
+	if err != nil {
+		return nil, err
+	}
+	if len(strings.TrimSpace(r.AirdropId)) == 0 {
+		return nil, ErrInvalidClaimAirdropId
+	}
+
+	return &ClaimPacketMetadata{
+		AirdropId:     r.AirdropId,
+		StrideAddress: address,
+		Enabled:       true,
```

Comment:
I'm finding the `Enabled` concept a bit confusing. How are you thinking about this?

Reply:
Yeah the variable name is pretty bad. The idea is that each packet is either just a normal packet or it's a packet intended for claim/liquid stake forwarding. 

To me, having an abstraction make sense:
* Grab the packet
* Inspect it to see if there's any fowarding specifications
* Store that info in a struct so that later you can easily check it with an if statement as what's done in module_ibc

That said, maybe some better name options would be `RoutingEnabled`, `Forward`, `RoutingTriggered`, etc.

Reply:
refactored